### PR TITLE
Flame Graph

### PR DESCRIPTION
**Pull request by adamsitnik**
This PR is recreated #440

PerfView is very powerful tool, but sometimes for some inexperienced users like me it is hard to get a fast overview of what takes most of the CPU if there are plenty of methods on the stack.

This PR provides a simple implementation of the "Flame Graph".

It is just different visualization of the "call tree" view.

Sample result (for the "tutorial.exe")

image

Sample result for more complicated app (dotnet cli):

image

1. I added new tab (before the "notes")
2. I did not use SVG (most common way of implementing flame graph) because I would have to add external dependency to the project ([link](https://social.msdn.microsoft.com/Forums/vstudio/en-US/22978b1c-be4d-475d-8a57-371f324b4697/does-wpf-support-svg?forum=wpf))
3. I decided to draw the rectangles/textblocks on a simple WPF Canvas. For small boxes, I use rectangles (too many textblocks on a single canvas = bad perf)
4. I have used some fancy color pallet to differentiate the boxes 
5. I added possibility to export the flame graph to png file (right click => save)
6. I added the docs
7. The canvas is redrawn on resize/when the tab gets the focus
8. Both tool tip and bottom panel contains method name, samples count and % of total time
9. Since the flame graph is using data from call tree view, any filter applied to call tree view works for flame graph as well

### CODE REVIEW

**Comment by vancem** [comment]
@adamsitnik  There is a failure in the debug test run.   I looked into it a bit and it seems likely that it is issue #354 causihg the issue.   

You need to update StackWindows to that the TestIncludeItemOnCallerCalleeTabCallerAsync test has the folloiwng 
```
        [WpfFact(Skip = "Failing with indexOutOfRange and Debug testing.   See issue https://github.com/Microsoft/perfview/issues/354")]
```

**Comment by vancem** [comment]
PerfView has a mechanism where all files needed at runtime are emedded into the EXE itself and extracted at first launch.    You added a new *.PNG file which needs to be emedded into PerfView.exe.

To do this modify src\PerfView\PerfVIew.csproj to add the following lines
```
    <EmbeddedResource Include="SupportFiles\Images\FlameGraphView.png">
      <Type>Non-Resx</Type>
      <WithCulture>false</WithCulture>
      <LogicalName>.\Images\FlameGraphView.png</LogicalName>
    </EmbeddedResource>
```
at line 417 (after all the other 'EmbeddedResource elements), but within the ItemGroup element.

**Comment by sharwell** [comment]
:memo: I restarted the build, which may pass on second try with no changes.

File: src/PerfView/FlameGraph.cs [hunk 1]
```diff
@@ -0,0 +1,160 @@
+﻿using System;
```
Comment (vancem) [inline src/PerfView/FlameGraph.cs]: Please put the FlameGraph.cs file in the 'src\perfView\StackViewer' directory.

File: src/PerfView/SupportFiles/UsersGuide.htm [hunk 2]
```diff
@@ -3246,6 +3247,21 @@ <h3>
...
     <!--  ****************** -->
+    <h3>
+        <a id="FlameGraphView">Flame Graph View</a>
+    </h3>
+    <p>
+        The flame graph view shows the same data as call tree view, but using different visualization.&nbsp;&nbsp;&nbsp;
+        It gives you very intelligible overview.&nbsp;&nbsp;
+        The graph starts at the bottom. Each box represents a method in the stack. Every parent is the caller, children are the callees.
+        The wider the box, the more time it was on-CPU. The samples count is shown in the tooltip and in the bottom panel.
+        To change the content of the flame graph you need to apply the filters for call tree view.&nbsp;&nbsp;
```
Comment (vancem) [inline src/PerfView/SupportFiles/UsersGuide.htm]: Can you put a hyperlink to some documentation on how to use FlameGraphs that already exists somewhere on the internet?

**Comment by vancem** [comment]
Please update the perfView version number (in the Directory.Build.Props file (it is in the 'Solution Items' folder in the solution).  to be 2.0.1 and then udpate the 'ReleaseNotes' section of the UsersGuid.htm to indicate that flame graphs have been added in this version of PerfView).

File: src/PerfView/StackViewer/StackWindow.xaml.cs [hunk 3]
```diff
@@ -411,6 +412,8 @@ public void SetStackSource(StackSource newSource, Action onComplete = null)
                             cumMax / 1000000, controller.GetStartTimeForBucket((HistogramCharacterIndex)cumMaxIdx));
                     }
 
+                    RedrawFlameGraph();
```
Comment (vancem) [inline src/PerfView/StackViewer/StackWindow.xaml.cs]: I would like this to be lazy.   Simply have a bit that marks that the flame graph is invalid at this point.  Then have an event when the canvas becomes visible that actually calls the RedrawFlamGraph.

This avoids a bunch of work that may never be used (if users never look at the flamegraph.

**Comment by vancem** [comment]
Note that I don't actually get the assert locally, I get the IndexOutOfRange exception (however maybe that exception is causing the cascade error).  

Here is where that 
```
  System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
   at System.ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument argument, ExceptionResource resource)
System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
   at System.ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument argument, ExceptionResource resource)
   at System.Collections.Generic.List`1.System.Collections.IList.get_Item(Int32 index)
   at System.Windows.Data.ListCollectionView.InternalItemAt(Int32 index)
   at System.Windows.Data.ListCollectionView.GetItemAt(Int32 index)
   at System.Windows.Controls.ItemCollection.GetItemAt(Int32 index)
   at System.Windows.Controls.ItemCollection.get_Item(Int32 index)
   at PerfViewTests.StackViewer.StackWindowTests.<>c__DisplayClass8_0.<<TestIncludeItemAsync>b__2>d.MoveNext() in G:\repos\perfView\src\PerfView.Tests\StackViewer\StackWindowTests.cs:line 107
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
   at System.ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument argument, ExceptionResource resource)
   at System.Collections.Generic.List`1.System.Collections.IList.get_Item(Int32 index)
   at System.Windows.Data.ListCollectionView.InternalItemAt(Int32 index)
   at System.Windows.Data.ListCollectionView.GetItemAt(Int32 index)
   at System.Windows.Controls.ItemCollection.GetItemAt(Int32 index)
   at System.Windows.Controls.ItemCollection.get_Item(Int32 index)
   at PerfViewTests.StackViewer.StackWindowTests.<>c__DisplayClass8_0.<<TestIncludeItemAsync>b__2>d.MoveNext() in G:\repos\perfView\src\PerfView.Tests\StackViewer\StackWindowTests.cs:line 107
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
   at PerfViewTests.Utilities.PerfViewTestBase.<>c__DisplayClass9_0`1.<<RunUITestAsync>b__1>d.MoveNext() in G:\repos\perfView\src\PerfView.Tests\Utilities\PerfViewTestBase.cs:line 72
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
   at System.ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument argument, ExceptionResource resource)
   at System.Collections.Generic.List`1.System.Collections.IList.get_Item(Int32 index)
   at System.Windows.Data.ListCollectionView.InternalItemAt(Int32 index)
   at System.Windows.Data.ListCollectionView.GetItemAt(Int32 index)
   at System.Windows.Controls.ItemCollection.GetItemAt(Int32 index)
   at System.Windows.Controls.ItemCollection.get_Item(Int32 index)
   at PerfViewTests.StackViewer.StackWindowTests.<>c__DisplayClass8_0.<<TestIncludeItemAsync>b__2>d.MoveNext() in G:\repos\perfView\src\PerfView.Tests\StackViewer\StackWindowTests.cs:line 107
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
   at PerfViewTests.Utilities.PerfViewTestBase.<>c__DisplayClass9_0`1.<<RunUITestAsync>b__1>d.MoveNext() in G:\repos\perfView\src\PerfView.Tests\Utilities\PerfViewTestBase.cs:line 72
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at PerfViewTests.Utilities.PerfViewTestBase.<>c__DisplayClass9_0`1.<<RunUITestAsync>b__1>d.MoveNext() in G:\repos\perfView\src\PerfView.Tests\Utilities\PerfViewTestBase.cs:line 77
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
```

I have also captured a trace of running just this test at \\clrmain\public\writable\users\vancem\PerfViewTestFailure.etl.zip 

It shows an ArgumentOutOfRange exception at  the line
```
 object selectedItem = dataGrid.Grid.Items[0];
```
This is what I see in the debugger as well.  In the debugger the 'Items' list has a count of 0.

**Comment by adamsitnik** [comment]
@vancem Done! Good idea with redrawing the graph only when it is visible.

**Comment by vancem** [comment]
Looks good to me

**Comment by goldshtn** [comment]
🎊  Very happy to see this merged! Nowadays I find that flame graphs are more useful than stack trees or other forms of stack visualization in 99% of the cases. Minor nitpick: if we could get rid of the "module" noise so that as much as possible of the actual function name would be shown in the flame rectangles, it would be even more awesome...

**Comment by adamsitnik** [comment]
>  Minor nitpick: if we could get rid of the "module" noise so that as much as possible of the actual function name would be shown in the flame rectangles, it would be even more awesome...

ok. 

Once you start using it and have some more comments please create a separate issue. I would like to make the Flame Graphs user-friendly. And I did not forget about the colors, I just did not figure this out yet.

**Comment by xiangzhai** [comment]
:mips-interest

**Pull request merged**